### DISCUSSION TITLE
Proficiencies: Future directions

### DISCUSSION THREAD

**Discussion by I-am-Erk** [CleverRaven/Cataclysm-DDA#49191] [Ideas]
This is a master thread for future development of proficiencies, where people can weigh in with discussion and ideas as well as link to specific issues.

I will split discussion areas by the various things that require proficiency support, and might need further code for it.

#### Fail states 
The first barrier I have found for cooking proficiencies is that crafting is mainly all-or-nothing. Lacking proficiency in most cooking related skills doesn't add all that much to prep time, or all that much to failure rate in terms of not being able to eat the product. Instead it makes it likely that what you'll make at the end won't taste good.

The best solution here would be to add [item quality levels](https://github.com/CleverRaven/Cataclysm-DDA/issues/35533) in [some form](https://github.com/CleverRaven/Cataclysm-DDA/issues/44740), but that's an enormous undertaking.

Another more feasible option might be to add a tag `[yucky]`, similar to `[mushy]`, which preserves nutritional data but overrides Fun. I think that's probably more achievable. This behaviour should be specified in the *recipe* somewhere, so that things that can't really be made yucky can't get it (boiled water eg). `[dry]` might be another option, reducing enjoyability and quench.

#### Substitutions and tricky ingredients
Another barrier I've run into is that we allow all kinds of wacky substitutions with complex foraged ingredients. Really these should make it harder to prepare a recipe, but short of adding new recipes we currently don't have support for that.

One possibly slick way to do this, although a problem on the UX end, would be to add the ability to specify a proficiency associated with certain ingredients in the `requirements` block. On the JSON end this might look like this:
```json
  {
    "id": "seasoning_mild",
    "type": "requirement",
    "//": "Non-salt seasonings you might add to a dish to flavor it. Should be about a teaspoon, or smallest serving",
    "components": [
      [
        [ "hot_sauce", 1 ],
        [ "pepper", 1 ],
        [ "seasoning_salt", 1 ],
        [ "seasoning_italian", 1 ],
        [ "garlic_clove", 1 ]
      ]
    ],
    "substitutions": [
      { "component": "wild_herbs", "quantity": 1, "proficiency": "foraged_edibles" },
      { "component": "soysauce", "quantity": 1, "proficiency": "scavenger_cooking" }
    ]
  }
```
In this model the original `components` syntax is preserved but things that would be harder to use can be then given a proficiency block. It would be hard to represent this in the UX, but perhaps we could do it by marking those ingredients with a `*` and footnoting `*: These ingredients are a difficult substitution`. Later on, when I change `fail_multiplier` to `skill_penalty` we could potentially have the ingredients list say something like `wild herbs (-1)*` and the footnote read `*: You are missing proficiencies for these ingredients and will have the listed penalty to your cooking skill`.

Another similar option here would be to specify the proficiency in the *item definition of the ingredient*, so eg. wild herbs would always require a forager cooking proficiency. This has the advantage of possibly being useful for tools with item qualities but the disadvantage of not recognizing that something like soy sauce is sometimes a totally appropriate seasoning, and sometimes a substitution.

If doable, this would allow a *lot* of other stuff outside cooking, and would generally be a big boon to improving the quality and feel of scavenging and repurposing stuff.

## Butchery/Tanning
As these use activation rather than recipes, a bit of C++ support is required to add them. It should not be difficult based on my experience implementing it for Devices.

I'm open to feedback on how to build a butchery proficiency tree. I don't know enough about it to suggest a good one but perhaps something like:
```
Principles of Butchery
 -- Field Butchery - for quartering/field prepping corpses
 -- Skinning
 -- Skilled Butchery - gives a speed and yield boost
 -- Survival Butchery - Practiced when using low-end butcher tools, reduces penalties for having lower tool levels
```

## Combat/Weapons

More to come here. Proficiencies should be able to adjust your rate of stamina loss, reduce strength requirements for weapons, and affect your movement through particular terrains, for example.

### Archery
Archery has some of the clearest proficiency related benefits. To implement these best we would want to have better modelling of muscle fatigue.
```
Basic Archery Form: Reduces the stamina (or muscle fatigue) cost of archery significantly. The base rate might need increasing for this to work.
 -- Expert Archery Form: Reduces the strength requirement for bows by 1.
  -- Master Archery Form: Reduces the strength requirement for bows by 1 more.
```

## Athletics
Athletics proficiencies should add movement modes or reduce their difficulty, or replicate some things currently represented by traits.

### Parkour
This trait should be nerfed and made into a couple proficiencies.

- Basic Parkour - cannot be learned automatically, has to be taught or taken at creation. Gives current benefits but does not apply to natural objects - or at least things with the "FLOWER", "ORGANIC", "SHRUB", "FUNGUS", or "DIGGABLE" tags, which covers almost everything where parkour doesn't make sense (how do you parkour through tall grass?)
- Parkour Expert - requires Basic Parkour, can be learned through practice. Somewhat reduces fall damage.
- Trailblazer - identical to basic parkour but applies to everything Parkour now doesn't.

### Climbing
This is a full issue into itself, but basically we need to add an athletics check to climbing things like walls and trees, with fail states of "takes longer and costs more stamina" or "you fall and get hurt". Then we can add some climbing proficiencies:

- Tree Climbing: Dramatically reduces the 'fall and get hurt' fail rate for trees.
- Wall Climbing: As above for walls.
- Wall Jumping: Requires Basic Parkour and Wall Climbing, increases the speed you climb walls (but does not reduce stamina cost, you lose more stamina per turn to, in the end, spend the same amount of stamina you would have without the proficiency)

### Foraging
There are several parts of survival that need proficiency, but one that merits some thought is foraging. In specific, foraging is something that one can't necessarily learn by trial-and-error in the timespans we're talking about in play.

I suggest we split foraging into two basic categories: things that can kill you or make you sick if you mess up, and things that can't. The more dangerous things should be in proficiencies that cannot be learned without assistance.
- On that note, we should probably either make it possible to learn a `can_learn = false` proficiency if you have a reference book, or add a `can_learn_book = true` parameter.
- Proficiencies should not have a role in harvesting very common edible fruits like blackberries, strawberries, etc.
- We might want to consider breaking each proficiency into seasonal variants

Suggested foraging proficiencies:
- Common Wild Plants: This adds a yield and time bonus to harvesting burdock, dandelion, etc that are easy to harvest but not as widely known as edible. `can_learn = true`.
- Common Plant Lookalikes: For fiddleheads, wild carrot, and other plants that can be mistaken for similar plants until you know the identifying traits. Should be easy to learn but `can_learn = false`. Without this proficiency, harvesting should not be an option, at least until we can have accidental poisoning.
- Difficult Plant Lookalikes: As above, but this would be for plants you can't forage unless you really know what you're doing. I don't know if we have any of these yet.
- Medicinal Plants: For foraging spurge and other things with medicinal role that are not widely known. `can_learn = false`. Without this proficiency, harvesting should not be an option.
- Foraging from Trees: for getting tanbark, cambion, and other things. `can_learn = true`. Skill improves yield and adds a time bonus.
- Mushroom foraging: `can_learn = false`, having this would enable mushroom harvesting, perhaps from underbrush. We'd have to add mushrooms. We may wish a 'common' and 'difficult' option with this as well, enabling fancier mushrooms if you know how to get them.

**Comment by big-smitty** [comment]
Imo, the proficiency system needs to use prerequisites and not just lump all proficiencies in one batch. 
The current system leads to some problems:
![image](https://user-images.githubusercontent.com/85451361/120932123-6cad4700-c6a9-11eb-95ae-800b278e705a.png)
Here we have to know "principles of leatherworking" as well as just "leatherworking".  Same deal with "principles of welding" and "welding".
A combined chance of failure 781.3x more likely (!!!) is totally unbalanced.  It would make more sense to combine some proficiencies, and disregard advanced proficiencies if you are lacking their prerequisite.

**Reply by I-am-Erk** [reply]
You're using an older experimental there.

Edit: sorry, hit send early.

Failure rates are lower now, and it is entirely intentional behaviour that failure and time multipliers escalate dramatically if you are missing too many proficiencies.

**Reply by big-smitty** [reply]
Oh, awesome. I didn't see that it was changed in newer versions. Thanks

**Comment by Rabidweasel** [comment]
Combat proficiencies seems like an interesting concept, could this be a replacement / enhancement to the existing martial arts feature?

Rather than learning a martial art all in one go and then having set abilities unlocked at a given skill level, characters could gain proficiency in specific individual tactics and techniques through training and practice. This would theoretically open the door to being able to organically combine techniques from multiple arts. I suppose that for balance and plausibility's sake there would need to be a way of making certain arts restricted which could be done by having some techniques require a particular stance, similar to how the current system works.

**Reply by I-am-Erk** [reply]
Eventually they will need to combine with martial arts, but I think it's going to be more a matter of treating martial arts as proficiencies and perhaps splitting them into a few levels (beginner tae kwon do, advanced tae kwon do, tae kwon do master), rather than fully breaking them up like that.

**Comment by 3l3m01** [comment]
For Foraging I think it wold make more sense if you can harvest everything, but without the right proficiencies you get an unknown herb. This would probably need some C++ changes to implement, but could also be made into a more generic "Identification system".

Basically there would be a dummy items for each real item, with the same name per category (Unknown herb, Unknown fruit, Unknown berry, etc). 
Some values are the same as the original (You can eat anything... once... ), but some, like the name and description are overwritten. 
The new "fake" items must also know what they are identified into (this could be several things too, but that might be problematic with things that can drop from multiple places, with different chances). 
You cannot craft with unknown item, and cannot Use them. New, bad items (mostly poisonous stuff) must be added to make Foraging without the right know-how risky.

You auto identify things if you have/get the right proficiency, or you can get an NPC to identify for you.

Different fake items would not stack together (I think), but I'm not sure how big a problem that is.

**Reply by 3l3m01** [reply]
Instead of making a fake item, it might be better to make the game able to "lie" to the player of the inventory screens if an item is unidentified, and make changes (like what to change the name and description into, and what proficiency you need to identify it) to the original item.

**Reply by pehamm** [reply]
We already have this with mushrooms, only that they use skills instead of proficiencies. You can find normal, poisonous or hallucinogenic mushrooms, but with missing skills they just appear as mushrooms. You can cook and craft with them, but if you cook with them the results can still be poisonous or make you hallucinate. Changing this from skill checks to proficiency checks should be pretty simple.

**Reply by I-am-Erk** [reply]
Oh, did someone actually implement that with mushrooms? We've been talking about that for ages. That means we should also implement different grades of meth that all look the same, soon.

EDIT: ah, no, it's the old problem for mushrooms where it's hard coded, that's been the barrier for a while. We would still need to jsonize the ability to list unidentified plants to allow this.

The thing here is that we can fairly easily implement a "you don't know how to harvest that yet" json field, but adding the infrastructure for randomly poisoning yourself is harder and not strictly necessary. Also, the 'you can't even try' thing is just the right level of annoying to get the job done but also encourage people to add the needed infrastructure for the better system. And then we can also add proficiency xp eventually for eating things you scavenged.

**Comment by CivilisationDDA** [comment]
I'm not sure whether I should post these links here but there has been some discussion of proficiencies for vehicle use here: #49018 and mentioned by core developer @mlangsdorf here: https://discourse.cataclysmdda.org/t/helicopters-how-to/23168/18

**Comment by SleepyJoeAZ** [comment]
I really wish there was a high end cap or at least diminishing returns on the failure rate and time increase.

**Reply by I-am-Erk** [reply]
If I were to add a high end cap, it would be "when your multiplier reaches *x* you can no longer attempt this recipe". There isn't a diminishing returns where lack of knowledge and understanding starts to have less impact because you're missing even more knowledge and understanding.

**Reply by SleepyJoeAZ** [reply]
Well i have the electronics skill and i cant craft an electrohack because i cant find the 101 book for the integrated circuits practice thing. So basically im SOL until i find that random ass book.

**Reply by I-am-Erk** [reply]
That kinda seems like the sort of thing you'd need a reference book to figure out. If the problem is with the rarity of the appropriate books that has pretty much nothing to do with proficiencies.

**Comment by coyo7e** [comment]
Archery currently weakens a character ludicrously quickly. This sucks since bows are now pretty usable however ending up wearied after a half-dozen shots is pretty rough. I like the idea of the proficiencies decreasing weariness drain by a lot since I've used bows a lot IRL and they aren't really that tiring unless you're totally out of shape and don't know what you're doing.. Another idea is that when firing a bow you could screw up badly enough to hurt yourself - that's why they make finger and arm guards for archers, the string slapping your inner forearm can be pretty painful, and you can also hurt your fingertips and get busted blood vessels in your fingers by pulling on a bow with a high pull strength, without building up calluses and learning to hold it.. This would also potentially allow a new weapon mod "mechanical bow release" which is a little handle with a trigger and hook, that a person loops around the string to keep from having to use their bare hands - and any bow could use that kind of mod. https://www.youtube.com/watch?v=6-p5eZhis2c

Food failures tasting bad is a good idea, although probably it ought to reduce the nutrition by a random amount (possibly even making it inedible if it's bad enough of a failure) and use up extra resources, or something similar. I've made some pretty bad stuff cooking IRL before, and usually it's edible but sometimes I mix up Tbsp and Tsp measures or burn something too badly, and it's just not consumable by anything but my dog. I can see being desperate and eating it anyway since we can force a character to eat raw organ meat, rotten corpses and other rotten foods etc. at a reduction in nutrition and happiness, or making the character ill or dehydrated etc.
